bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/DataClientVeneerApi.java

```diff
@@ -229,8 +249,97 @@ protected void onCompleteImpl() {
   }
 
   /** wraps {@link ServerStream} onto HBase {@link ResultScanner}. */
-  private static class RowResultScanner extends AbstractClientScanner {
+  static class PaginatedRowResultScanner extends AbstractClientScanner {
+    // Percentage of max number of rows allowed in the buffer
+    private static final double WATERMARK_PERCENTAGE = .1;
+    private static final int MIN_BYTE_BUFFER_SIZE = 100 * 1024 * 1024;
+    private static final double DEFAULT_BYTE_LIMIT_PERCENTAGE = .1;
+    private static final RowResultAdapter RESULT_ADAPTER = new RowResultAdapter();
 
+    private final Meter scannerResultMeter =
+        BigtableClientMetrics.meter(BigtableClientMetrics.MetricLevel.Info, "scanner.results");
+    private final Timer scannerResultTimer =
+        BigtableClientMetrics.timer(
+            BigtableClientMetrics.MetricLevel.Debug, "scanner.results.latency");
+
+    private ServerStream<Result> serverStream;
+    private ByteString lastSeenRowKey = ByteString.EMPTY;
+    private Boolean hasMore = true;
+    private final Queue<Result> buffer;
+    private final Query.QueryPaginator paginator;
+    private final paginatorFunction wrapper;
+    private final int refillSegmentWaterMark;
+
+    static long maxSegmentByteSize =
+        (long)
+            Math.max(
+                MIN_BYTE_BUFFER_SIZE,
+                (Runtime.getRuntime().totalMemory() * DEFAULT_BYTE_LIMIT_PERCENTAGE));
+    private long currentByteSize = 0;
+
+    PaginatedRowResultScanner(Query.QueryPaginator paginator, paginatorFunction wrapper) {
+      this.paginator = paginator;
+      this.wrapper = wrapper;
+      this.buffer = new ArrayDeque<>();
+      this.refillSegmentWaterMark =
+          (int) Math.max(1, paginator.getPageSize() * WATERMARK_PERCENTAGE);
+      this.serverStream = this.wrapper.func(this.paginator);
+    }
+
+    @Override
+    public Result next() {
+      try (Context ignored = scannerResultTimer.time()) {
+        if (this.buffer.size() < this.refillSegmentWaterMark
+            && this.serverStream == null
+            && hasMore) {
+          this.serverStream = this.wrapper.func(this.paginator);
+        }
+        if (this.buffer.isEmpty() && this.serverStream != null) {
+          this.waitReadRowsFuture();
+        }
+        scannerResultMeter.mark();
+        Result result = this.buffer.poll();
+        if (result != null) {
+          currentByteSize -= Result.getTotalSizeOfCells(result);
+        }
+        return result;
+      }
+    }
+
+    @Override
+    public void close() {
+      if (this.serverStream != null) {
+        this.serverStream.cancel();
+        this.serverStream = null;
+      }
+    }
+
+    public boolean renewLease() {
+      throw new UnsupportedOperationException("renewLease");
+    }
+
+    private void waitReadRowsFuture() {
+      Iterator<Result> iterator = this.serverStream.iterator();
```

Comment:
I think it would be significantly more efficient if you use the async api to fetch the next batch. ie

```java
SettableFuture<List<Result>> resultsFuture = SettableFuture.create();

dataClient.readRowsCallable(myRowMerger).call(req,  new ResponseObserver() {
   List<Result> results = new ArrayList();
   onStart(Controller c) {
     this.controller = c;
   }
   onResponse(Result r) {
      results.add(r);
      if (thresholdReached) {
         this.controller.cancel();
      }
   }
   onComplete() {
     resultsFuture.set(results);
   }
   onError(e) {
      resultsFuture.setExceptionally(e);
});

List<Result> results = resultsFuture.get();
....
```

Reply:
done